rfc/rfc004-distributed-document-format.md

```diff
@@ -0,0 +1,114 @@
+* **alg**: MUST be one of the following algorithms: `PS256`, `PS384`, `PS512`, `ES256`, `ES384` or `ES512`.
+  other algorithms SHALL NOT be used.
+* **cty**: MUST contain the type of the payload indicating how to interpret the payload encoded as string.
+* **crit** MUST contain the **sigt** and **version** headers.
+
+The **jku**, **jwk**, **kid** and **x5u** header parameters SHOULD NOT be used and MUST be ignored by when processing the document.
+
+In addition to the registered header parameters, the following header MUST be present as protected headers:
+* **sigt**: (signing time) MUST contain the signing time of the document in UTC as string, formatted according to [RFC3339](https://tools.ietf.org/html/rfc3339).
+* **version**: MUST contain the format version of the document as number. For this version of the format the version MUST be 1.
+* **prev**: MUST contain the reference (see section 3.2) of the preceding document, or a zeroed reference if a genesis document.
+
+The JWS payload MUST be the actual contents of the document. The format is unspecified so MAY any data type supported by the JSON including a nested JSON document.
+
+### 3.2. Document Reference
+The document reference uniquely identifies a document and is used to fill the **prev** header. It MUST be calculated by
+taking the bytes of the JWS EXACTLY as received and hashing it using SHA-1.
+
+When serializing a reference to string form it MUST be hexadecimal encoded and SHOULD be lowercase.
+
+Example:
+
+```148b3f9b46787220b1eeb0fc483776beef0c2b3e```
+
+### 3.3. Ordering
+Documents ordering is casual: a document MUST refer to the last known document on the network by specifying its
+reference in the *prev* field. This is essential for maintaining consistent order across peers.
+The *iat* field SHOULD have a timestamp equal to or after the referenced document's *iat* field.
+The first document published on the network (genesis document) MUST have a zeroed *prev* field:
+
+```0000000000000000000000000000000000000000```
+
+There MUST only be one genesis document for a network. Subsequent genesis documents MUST be ignored.
+
+TODO: What about branching?
+
+
+### 3.4. Validation
+Before interpreting a document's payload it SHOULD be validated according to the following rules:
+
+1. Assert that if it's a genesis document, we didn't already receive one.
+2. Assert that the previous document (*prev* field) is valid. Preceding documents should be validated first.
+3. Verify the document signature:
+   * Validate keyUsage, validity of the certificate in the *x5c* field and whether the issuer is trusted.
+   * Verify the cryptographic signature with the public key from the certificate.
+   * Assert that the certificate was valid at the time of signing (as specified by *iat*).
+   * Assert that the certificate was not revoked on at time of signing. 
+
+If any of the steps above fail the document SHOULD be rejected and its payload SHALL NOT be deemed valid.
```

Comment:
rejected means? discarded/ignored

Reply:
Good question, I'd say discarded and any documents after that in the chain. But on second thought a failure of one of the validation steps doesn't mean the rest of the chain's integrity or authenticity after the failed document is compromised. So if we don't process the rest of the chain, I'd say it's out of simplicity. Problem is also that the validation is not deterministic: when a certificate is revoked later on it would invalidate not just this particular document but also any documents after it.